server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseSetMethod.scala

```diff
@@ -0,0 +1,84 @@
+import org.apache.james.mailbox.MailboxSession
+import org.apache.james.metrics.api.MetricFactory
+import org.apache.james.util.date.ZonedDateTimeProvider
+import org.reactivestreams.Publisher
+import play.api.libs.json.{JsError, JsObject, JsSuccess}
+import reactor.core.scala.publisher.SMono
+
+class VacationResponseSetMethod @Inject() (serializer: Serializer,
+                                           vacationRepository: VacationRepository,
+                                           zonedDateTimeProvider: ZonedDateTimeProvider,
+                                           metricFactory: MetricFactory) extends Method {
+  override val methodName: MethodName = MethodName("VacationResponse/set")
+
+  override def process(capabilities: Set[CapabilityIdentifier],
+                       invocation: Invocation,
+                       mailboxSession: MailboxSession,
+                       processingContext: ProcessingContext): Publisher[Invocation] = {
+    asVacationResponseSetRequest(invocation.arguments)
+      .flatMap(vacationResponseSetRequest => update(mailboxSession, vacationResponseSetRequest)
+        .map(updateResult => createResponse(invocation, vacationResponseSetRequest, updateResult)))
+  }
+
+  private def update(mailboxSession: MailboxSession, vacationResponseSetRequest: VacationResponseSetRequest): SMono[UpdateResult] = {
+    vacationResponseSetRequest.parsePatch()
+        .fold(e => SMono.just(UpdateFailure(e)),
+          patch => SMono.fromPublisher(vacationRepository.modifyVacation(toVacationAccountId(mailboxSession), patch.asVacationPatch))
+            .`then`(SMono.just(UpdateSuccess())))
+  }
+
+  private def toVacationAccountId(mailboxSession: MailboxSession): org.apache.james.jmap.api.vacation.AccountId = {
+    org.apache.james.jmap.api.vacation.AccountId.fromUsername(mailboxSession.getUser)
+  }
+
+  private def asVacationResponseSetRequest(arguments: Arguments): SMono[VacationResponseSetRequest] = {
+    serializer.deserializeVacationResponseSetRequest(arguments.value) match {
+      case JsSuccess(vacationResponseSetRequest, _) => SMono.just(vacationResponseSetRequest)
+      case errors: JsError => SMono.raiseError(new IllegalArgumentException(serializer.serialize(errors).toString))
+    }
+  }
+
+  private def createResponse(invocation: Invocation,
+                             vacationResponseSetRequest: VacationResponseSetRequest,
+                             updateResult: UpdateResult): Invocation = {
+    val response = VacationResponseSetResponse(
+      accountId = vacationResponseSetRequest.accountId,
+      newState = State.INSTANCE,
+      updated = updateResult.updated,
+      notUpdated = updateResult.notUpdated)
+
+    Invocation(methodName,
+      Arguments(serializer.serialize(response).as[JsObject]),
+      invocation.methodCallId)
+  }
+
+  sealed trait UpdateResult {
+    def updated: Option[Map[String, VacationResponseUpdateResponse]]
```

Comment:
What do you think of it to be just an option of VacationResponseUpdateResponse, and the jsobject with a "singleton" id be handled in the json writer?

Reply:
I guess it is so that the JSON serialization is Straightforward

Reply:
For the option, if I ask `"update": ["notTheRightId"]` then updated is empty.

The filtering trick to not return an empty updated relies on an optional typing.